{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "077e981e",
   "metadata": {},
   "source": [
    "## Clases en Python\n",
    "\n",
    "Se trata de una forma de estructurar objetos como si se tratara de un mapa arquitectónico que determina las características o propiedades de los objetos que obtenemos de usarla."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca31588b",
   "metadata": {},
   "source": [
    "### Ejercicio 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "44991193",
   "metadata": {},
   "outputs": [],
   "source": [
    "from numpy.random import permutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "dd05339d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5 14.008 239.85000000000002\n"
     ]
    }
   ],
   "source": [
    "class mamifero():\n",
    "    # Atributos de clase\n",
    "    vertebrado = True\n",
    "    amamanta = True\n",
    "\n",
    "    # Atributos de instancia\n",
    "    def __init__(self, alimentacion: str, altura: float, peso: float, descendencia: int = 0):\n",
    "        self.alimentacion = alimentacion\n",
    "        self.altura = altura\n",
    "        self.peso = peso\n",
    "        self.descendencia = descendencia\n",
    "    \n",
    "    def reproducirse(self, max_progenie: int):\n",
    "        self.descendencia += permutation(max_progenie)[0]\n",
    "\n",
    "    def crecer(self, crecimiento):\n",
    "        self.altura += crecimiento\n",
    "        self.peso += crecimiento * 0.4\n",
    "        return self.peso\n",
    "\n",
    "perro = mamifero(alimentacion='carnívoro', altura=2.008, peso=235.05)\n",
    "perro.reproducirse(6)\n",
    "perro.crecer(12)\n",
    "print(perro.descendencia, perro.altura, perro.peso)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cf94178",
   "metadata": {},
   "source": [
    "# Herencia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "63f52435",
   "metadata": {},
   "outputs": [],
   "source": [
    "class monotrema(mamifero):\n",
    "    espolon = True\n",
    "\n",
    "    def poner_huevos(self, max_huevos):\n",
    "        huevos = 0        \n",
    "        for n in range(max_huevos):\n",
    "            if permutation(2)[0]:\n",
    "                huevos += 1\n",
    "        if huevos > 0:\n",
    "            self.reproducirse(huevos)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5990a495",
   "metadata": {},
   "source": [
    "## Ejericio 3. Práctica de Herencia\n",
    "\n",
    "De la clase `gen` habremos de generar una subclase `tRNA` y una subclase de la anterior denominada `RNA no codificante`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1076805f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GC Content: 56.25%, 2 frame: ['TGA', 'AAG', 'CGT', 'TAA', 'CGG', 'CCA', 'GGC', 'AAC', 'AAG', 'AGG', 'TGT', 'TTG', 'ATC', 'TCA', 'TCC', 'GTG', 'ATC', 'ACA', 'TCA', 'GCC', 'AGA', 'CAG', 'GTA', 'TGC', 'CGC', 'CGA', 'CGC', 'GTG', 'CGG', 'AAA', 'TCG', 'CGC', 'AGC', 'GTT', 'TGG', 'GGT', 'TCC']\n"
     ]
    }
   ],
   "source": [
    "class gene():\n",
    "    \n",
    "    def __init__(\n",
    "            self, \n",
    "            name: str, \n",
    "            seq: str = '', \n",
    "            gc_con: float = 0.0, \n",
    "            organism: str = '', \n",
    "            lecture_frame: int = 1, \n",
    "    ):\n",
    "        self.name = name\n",
    "        self.length = 0\n",
    "        self.seq = seq.upper()\n",
    "        self.gc_percentage = gc_con\n",
    "        self.organism = organism\n",
    "        self.frame = lecture_frame\n",
    "        self.nt_type = ''\n",
    "\n",
    "    def gc_content(self) -> None:\n",
    "        total_gc = self.seq.count('G') + self.seq.count('C')\n",
    "        total = total_gc + self.seq.count('A') + self.seq.count('T')\n",
    "        self.gc_percentage = ((total_gc / total) * 100) if total > 0 else None\n",
    "    \n",
    "    def seq_length(self) -> int:\n",
    "        self.length = len(self.seq)\n",
    "        return self.length\n",
    "\n",
    "    def first_frame(self):\n",
    "        frame_codons = []\n",
    "        for i in range(0,len(self.seq),3):\n",
    "            codon = self.seq[i:i+3]\n",
    "            if len(codon) == 3:\n",
    "                frame_codons.append(codon)\n",
    "        return frame_codons\n",
    "    \n",
    "    def second_frame(self):\n",
    "        frame_codons = []\n",
    "        for i in range(1,len(self.seq),3):\n",
    "            codon = self.seq[i:i+3]\n",
    "            if len(codon) == 3:\n",
    "                frame_codons.append(codon)\n",
    "        return frame_codons\n",
    "    \n",
    "    def third_frame(self):\n",
    "        frame_codons = []\n",
    "        for i in range(2,len(self.seq),3):\n",
    "            codon = self.seq[i:i+3]\n",
    "            if len(codon) == 3:\n",
    "                frame_codons.append(codon)\n",
    "        return frame_codons\n",
    "    \n",
    "    def get_frame_codons(self):\n",
    "        match self.frame:\n",
    "            case 1:\n",
    "                return self.first_frame()\n",
    "            case 2:\n",
    "                return self.second_frame()\n",
    "            case 3:\n",
    "                return self.third_frame()\n",
    "            case _:\n",
    "                raise ValueError('\\n[ERROR] No such frame have been found.\\n')\n",
    "    \n",
    "    def base_type(self) -> None:\n",
    "        self.nt_type = 'RNA' if 'U' in self.seq else 'DNA'\n",
    "\n",
    "lexA = gene(\n",
    "    name='lexA',\n",
    "    seq='ATGAAAGCGTTAACGGCCAGGCAACAAGAGGTGTTTGATCTCATCCGTGATCACATCAGCCAGACAGGTATGCCGCCGACGCGTGCGGAAATCGCGCAGCGTTTGGGGTTCC',\n",
    "    organism='Escherichia coli K-12',\n",
    "    lecture_frame=2\n",
    ")\n",
    "lexA.gc_content()\n",
    "print(f'GC Content: {lexA.gc_percentage}%, {lexA.frame} frame: {lexA.second_frame()}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f814053",
   "metadata": {},
   "source": [
    "# Polimorfismo\n",
    "\n",
    "Se trata de la supersición de métodos entre clases _emparentadas_, de tal forma que lo que hace un método en la clase padre sea más preciso o particular en la clase hija.\n",
    "\n",
    "#### Ejemplo\n",
    "\n",
    "La clase suma en Python, con el signo `+`, por un lado puede sumar números ya sean `int` o `float` o, bajo otro uso, juntar dos caracteres o cadenas en una única cadena."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9cc88f60",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'alimentacion': 'carnívoro',\n",
       " 'altura': 200,\n",
       " 'peso': 41000,\n",
       " 'descendencia': 0,\n",
       " 'acuatico': True}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class placentario(mamifero):\n",
    "\n",
    "    def __init__(self, alimentacion, altura, peso, acuatico: bool,descendencia: int = 0):\n",
    "        super().__init__(alimentacion, altura, peso, descendencia)\n",
    "        self.acuatico = acuatico\n",
    "\n",
    "    def crecer(self, crecimiento): # Ejemplo de polimorfismo --> overriding\n",
    "        self.altura += crecimiento\n",
    "        self.peso += crecimiento * 2\n",
    "        return self.peso\n",
    "\n",
    "cachalote = placentario('carnívoro', 200, 41*10**3, True)\n",
    "cachalote.__dict__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a7ad0dd2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41040 3.6399999999999997 "
     ]
    }
   ],
   "source": [
    "class marsupial(mamifero):\n",
    "\n",
    "    def __init__(self, alimentacion, altura, peso, descendencia = 0):\n",
    "        super().__init__(alimentacion, altura, peso, descendencia)\n",
    "\n",
    "    def crecer(self, crecimiento):\n",
    "        return super().crecer(crecimiento) * 0.7\n",
    "    \n",
    "tlacuache = marsupial('omnívoro', 30, 1.2)\n",
    "\n",
    "for mamiferito in (cachalote, tlacuache):\n",
    "    print(mamiferito.crecer(10), end=' ')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b2d5f80",
   "metadata": {},
   "source": [
    "# Pruebas de los ejercicios 3 y 4\n",
    "\n",
    "Este bloque será dedicado a las pruebas de las clases y subclases previo a mayores modificaciones."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dd913fff",
   "metadata": {},
   "outputs": [],
   "source": [
    "class tRNA(gene):\n",
    "\n",
    "    CODON_TO_AA = {\n",
    "        'AUG': 'Met',\n",
    "        'UUU': 'Phe', 'UUC': 'Phe',\n",
    "        'UUA': 'Leu', 'UUG': 'Leu', 'CUU': 'Leu', 'CUC': 'Leu', 'CUA': 'Leu', 'CUG': 'Leu',\n",
    "        'AUU': 'Ile', 'AUC': 'Ile', 'AUA': 'Ile',\n",
    "        'GUU': 'Val', 'GUC': 'Val', 'GUA': 'Val', 'GUG': 'Val',\n",
    "        'UCU': 'Ser', 'UCC': 'Ser', 'UCA': 'Ser', 'UCG': 'Ser', 'AGU': 'Ser', 'AGC': 'Ser',\n",
    "        'CCU': 'Pro', 'CCC': 'Pro', 'CCA': 'Pro', 'CCG': 'Pro',\n",
    "        'ACU': 'Thr', 'ACC': 'Thr', 'ACA': 'Thr', 'ACG': 'Thr',\n",
    "        'GCU': 'Ala', 'GCC': 'Ala', 'GCA': 'Ala', 'GCG': 'Ala',\n",
    "        'UAU': 'Tyr', 'UAC': 'Tyr',\n",
    "        'CAU': 'His', 'CAC': 'His',\n",
    "        'CAA': 'Gln', 'CAG': 'Gln',\n",
    "        'AAU': 'Asn', 'AAC': 'Asn',\n",
    "        'AAA': 'Lys', 'AAG': 'Lys',\n",
    "        'GAU': 'Asp', 'GAC': 'Asp',\n",
    "        'GAA': 'Glu', 'GAG': 'Glu',\n",
    "        'UGU': 'Cys', 'UGC': 'Cys',\n",
    "        'UGG': 'Trp',\n",
    "        'CGU': 'Arg', 'CGC': 'Arg', 'CGA': 'Arg', 'CGG': 'Arg', 'AGA': 'Arg', 'AGG': 'Arg',\n",
    "        'GGU': 'Gly', 'GGC': 'Gly', 'GGA': 'Gly', 'GGG': 'Gly',\n",
    "        'UAA': 'stop', 'UAG': 'stop', 'UGA': 'stop'\n",
    "    }\n",
    "\n",
    "    def __init__(\n",
    "            self, \n",
    "            name: str = '', \n",
    "            length: int = 0, \n",
    "            seq: str = '', \n",
    "            gc_con: float = 0.0, \n",
    "            lecture_frame: int = 1, \n",
    "            anticodon: str = 'CAU',\n",
    "            loops: bool = True,\n",
    "            discrim_base: str = 'A',\n",
    "            acceptor_stem: bool = True\n",
    "    ):\n",
    "        super().__init__(name=name, seq=seq, gc_con=gc_con,\n",
    "                 organism='', lecture_frame=lecture_frame)\n",
    "\n",
    "        self.anticodon = anticodon.upper().replace('T', 'U')\n",
    "        self.loops = loops\n",
    "        self.discrim_base = discrim_base\n",
    "        self.acceptor_stem = acceptor_stem\n",
    "        self.seq = self.seq.replace('T', 'U')\n",
    "    \n",
    "    def paired_codon(self) -> str:\n",
    "        comp = {\n",
    "            'A' : 'U',\n",
    "            'U' : 'A',\n",
    "            'C' : 'G',\n",
    "            'G' : 'C'\n",
    "        }\n",
    "        return ''.join(comp.get(b, b) for b in self.anticodon[::-1])\n",
    "    \n",
    "    def noncanonical_bases(self) -> bool:\n",
    "        allowed = {'A','U','C','G'}\n",
    "        return any(c not in allowed for c in self.anticodon)\n",
    "\n",
    "    def codon_translator(self) -> str:\n",
    "        if (l := len(anti := self.anticodon)) != 3:\n",
    "            raise ValueError(f\"\\n[ERROR] Anticodon must have 3 bases length, it has {l}\\n\")\n",
    "        \n",
    "        if self.noncanonical_bases():\n",
    "            raise ValueError(f'\\n[ERROR] The anticodon sequence {anti} contains noncanonical bases\\n')\n",
    "        \n",
    "        comp_codon = self.paired_codon()\n",
    "        aa = self.CODON_TO_AA.get(comp_codon)\n",
    "            \n",
    "        if not aa:\n",
    "            raise ValueError(f\"\\n[ERROR] No aminoacid could match {anti} anticodon\\n\")\n",
    "        return aa\n",
    "    \n",
    "class Protein(tRNA):\n",
    "    def __init__(\n",
    "        self, \n",
    "        name: str = '', \n",
    "        length: int = 0, \n",
    "        seq: str = '', \n",
    "        gc_con: float = 0.0, \n",
    "        lecture_frame: int = 1, \n",
    "        anticodon: str = 'CAU',\n",
    "        loops: bool = True,\n",
    "        discrim_base: str = 'A',\n",
    "        acceptor_stem: bool = True\n",
    "    ):\n",
    "        super().__init__(name=name, seq=seq, gc_con=gc_con,\n",
    "                 lecture_frame=lecture_frame, anticodon=anticodon,\n",
    "                 loops=loops, discrim_base=discrim_base, acceptor_stem=acceptor_stem)\n",
    "        self.aminoacids = ''\n",
    "        self.length = length\n",
    "\n",
    "    @staticmethod\n",
    "    def has_noncanonical(bases: str) -> bool:\n",
    "        allowed = {'A','U','C','G'}\n",
    "        return any(b not in allowed for b in bases.upper())\n",
    "\n",
    "    @staticmethod\n",
    "    def anticodon_method(anticodon: str) -> str:\n",
    "        comp = {\n",
    "            'A' : 'U',\n",
    "            'U' : 'A',\n",
    "            'C' : 'G',\n",
    "            'G' : 'C'\n",
    "        }\n",
    "        anti = anticodon.upper().replace('T','U')\n",
    "        return ''.join(comp.get(b, b) for b in anti[::-1])\n",
    "    \n",
    "    def codon_translator(self, anticodon: str) -> str: # Polymorphism\n",
    "        if (l := len(anti := anticodon)) != 3:\n",
    "            raise ValueError(f\"\\n[ERROR] Anticodon must have 3 bases length, it has {l}\\n\")\n",
    "        \n",
    "        if self.has_nocanonical(anti):\n",
    "            raise ValueError(f'\\n[ERROR] The anticodon sequence {anti} contains noncanonical bases\\n')\n",
    "        codon = self.anticodon_method(anti)\n",
    "        aa = self.CODON_TO_AA.get(anti, None)\n",
    "            \n",
    "        if not aa:\n",
    "            raise ValueError(f\"\\n[ERROR] No aminoacid could match {anti} anticodon\\n\")\n",
    "        return aa\n",
    "    \n",
    "    def codons_2_aa(self):\n",
    "        codons = self.first_frame()\n",
    "        aa_list = []\n",
    "        for c in codons:\n",
    "            aa = self.CODON_TO_AA.get(c)\n",
    "            if aa == 'stop':\n",
    "                break\n",
    "            if aa:\n",
    "                aa_list.append(aa)\n",
    "        self.aminoacids = '-'.join(aa_list)     \n",
    "\n",
    "    def seq_length(self) -> dict[str, int]:\n",
    "        nt_len = super().seq_length()\n",
    "        if not self.aminoacids:\n",
    "            self.codons_2_aa()\n",
    "        aa_len = len(self.aminoacids.split('-')) if self.aminoacids else 0\n",
    "        return {'Nucleotide seq': nt_len, 'Aminoacid seq': aa_len}\n",
    "    \n",
    "class ncRNA():\n",
    "    def __init__(self, seq: str = '', gc_con: float = 0, sec_struct: str = 'hairpin',\n",
    "                 cellular_loc: str = 'cytosol', seed_seq: str = None, complex: str = '',\n",
    "                 circular: bool = False):\n",
    "        self.length = len(seq)\n",
    "        self.gc_con = gc_con\n",
    "        self.seq = seq.upper().replace('T', 'U')\n",
    "        self.second_struct = sec_struct.lower()\n",
    "        self.cellular_loc = cellular_loc.lower()\n",
    "        self.circular = circular\n",
    "        self.seed_seq = seed_seq.upper().replace('T','U') if seed_seq else None\n",
    "        self.associated_complex = complex\n",
    "\n",
    "    def ncRNA_type(self) -> str:\n",
    "        if self.circular:\n",
    "            return 'circRNA'\n",
    "        elif self.length > 200:\n",
    "            return 'lncRNA'\n",
    "        elif self.seed_seq:\n",
    "            return 'miRNA'\n",
    "        elif 'PIWI' in self.associated_complex:\n",
    "            return 'piRNA'\n",
    "        else:\n",
    "            raise ValueError(f\"\\n[ERROR] No match for the given ncRNA\\n\")\n",
    "        \n",
    "    def seed_sequence(self, ncrna_type: str = ''):\n",
    "        if ncrna_type != 'miRNA':\n",
    "            return None\n",
    "        return self.seq[1:8]\n",
    "    \n",
    "    def gc_percentage(self) -> None:\n",
    "        total_gc = self.seq.count('G') + self.seq.count('C')\n",
    "        total = total_gc + self.seq.count('A') + self.seq.count('U')\n",
    "        self.gc_con = ((total_gc / total) * 100) if total > 0 else None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4964491a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene seq (DNA): ATGCAT\n",
      "tRNA seq (RNA): AUG\n",
      "tRNA anticodon: CAU\n",
      "Paired codon: AUG\n",
      "AA: Met\n",
      "Protein seq (RNA): AUGUUUUGA\n",
      "AA sequence: Met-Phe\n",
      "Lengths: {'Nucleotide seq': 9, 'Aminoacid seq': 2}\n",
      "ncRNA type: circRNA\n",
      "ncRNA GC%: 50.0\n"
     ]
    }
   ],
   "source": [
    "# Gene debe quedar con T\n",
    "gene_dna = gene(name=\"test_gene\", seq=\"ATGCAT\")\n",
    "print(\"Gene seq (DNA):\", gene_dna.seq)   # Debe mostrar con T\n",
    "assert 'T' in gene_dna.seq and 'U' not in gene_dna.seq\n",
    "\n",
    "# tRNA debe convertir a U\n",
    "trna = tRNA(seq=\"ATG\", anticodon=\"CAU\")\n",
    "print(\"tRNA seq (RNA):\", trna.seq)       # Debe mostrar con U\n",
    "assert 'U' in trna.seq and 'T' not in trna.seq\n",
    "\n",
    "# paired_codon: anticodón CAU → codón AUG → Met\n",
    "print(\"tRNA anticodon:\", trna.anticodon)\n",
    "print(\"Paired codon:\", trna.paired_codon())\n",
    "print(\"AA:\", trna.codon_translator())\n",
    "assert trna.paired_codon() == \"AUG\"\n",
    "assert trna.codon_translator() == \"Met\"\n",
    "\n",
    "# Protein hereda de tRNA, pero seq también está en U\n",
    "prot = Protein(seq=\"AUGUUUUGA\")  # AUG (Met), UUU (Phe), UGA (STOP)\n",
    "print(\"Protein seq (RNA):\", prot.seq)\n",
    "prot.codons_2_aa()\n",
    "print(\"AA sequence:\", prot.aminoacids)\n",
    "lengths = prot.seq_length()\n",
    "print(\"Lengths:\", lengths)\n",
    "assert prot.seq.startswith(\"AUG\")\n",
    "assert prot.aminoacids.startswith(\"Met\") or len(prot.aminoacids) > 0\n",
    "assert lengths[\"Nucleotide seq\"] == len(prot.seq)\n",
    "assert lengths[\"Aminoacid seq\"] > 0\n",
    "\n",
    "# ncRNA con semilla y circularidad\n",
    "ncrna = ncRNA(seq=\"AUGC\", circular=True)\n",
    "print(\"ncRNA type:\", ncrna.ncRNA_type())\n",
    "ncrna.gc_percentage()\n",
    "print(\"ncRNA GC%:\", ncrna.gc_con)\n",
    "assert ncrna.ncRNA_type() == \"circRNA\"\n",
    "assert round(ncrna.gc_con, 1) == 50.0\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mi_entorno",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
